In [1]:
import pandas as pd
import numpy as np

In [3]:
acci_df = pd.read_csv('/content/drive/MyDrive/cloud_ai/프로젝트/모듈1/데이터/교통사고데이터(2019-2021).csv',encoding='cp949')
acci_df.head(2)

/usr/local/lib/python3.8/dist-packages/IPython/core/interactiveshell.py:3326: DtypeWarning: Columns (57) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


acdnt_no  engn_code  acdnt_year    acdnt_dd_dc  dfk_code dfk_dc  \
0  2019010100100001          0        2019  2019년  1월  1일         2    화요일   
1  2019010100100002          0        2019  2019년  1월  1일         2    화요일   

   tmzon_div_1_code tmzon_div_1_dc  occrrnc_time_code occrrnc_time_dc  ...  \
0                 2             야간                  0             00시  ...   
1                 2             야간                  0             00시  ...   

   rn rde_id  rde_nm searchCondition  pageIndex recordCountPerPage  \
0 NaN    NaN     NaN             NaN          1                 10   
1 NaN    NaN     NaN             NaN          1                 10   

   searchConditionText searchKeyword  pageUnit zoneYn  
0                  NaN           NaN         3  False  
1                  NaN           NaN         3  False  

[2 rows x 89 columns]

# 1. 컬럼 제거

- 총 컬럼이 89개
- 유의미한 컬럼만 남기기 위해 각 컬럼의 의미 파악
  - 1) 의미가 불분명한 컬럼들 중 필요없는 컬럼을 걸러내기 위해 unique 함수를 사용해 추가 정제
    - 모두 같은 값으로 들어가있는 컬럼이 다수 존재
  - 2) 같은 값을 의미하는(ex: 운전자 연령-운전자 연령 코드) 컬럼들 중 의미파악이 용이한 컬럼만 남기고 나머지 한 컬럼은 삭제


In [ ]:
# 의미는 불분명하지만 그 값이 모두 같아서 분석에 필요없는 컬럼 파악
print(acci_df['acdnt_des'].unique(),
      acci_df['acdnt_frm_lv1'].unique(),
      acci_df['acdnt_frm_lv2'].unique(),
      acci_df['acdnt_frm_lv3'].unique(),
      acci_df['acdnt_sta_lv1'].unique(),
      acci_df['acdnt_sta_lv2'].unique(),
      acci_df['city_idt_code'].unique(),
      acci_df['engn_code'].unique(),
      acci_df['rn'].unique(),
      acci_df['spt_otlnmap_at'].unique(),
      acci_df['zoneYn'].unique(),
      sep='\n')


[nan]
[nan]
[nan]
[nan]
[nan]
[nan]
[nan]
[0]
[nan]
['없음']
[False]


In [4]:
# 필요한 컬럼만 남기고 전부 삭제
acci_df = acci_df[['acdnt_age_1_code','acdnt_age_2_code',
                  'acdnt_dc','acdnt_dd_dc',
                  'acdnt_gae_dc','acdnt_hdc','acdnt_mdc',
                  'acdnt_no','acdnt_pos',
                  'acdnt_year',
                  'bdy_injury_part_1_dc','bdy_injury_part_2_dc',
                  'dfk_dc','dmge_vhcle_asort_dc','dprs_cnt',
                  'hhdgw_at', 'inj_aplcnt_cnt',
                  'injury_dgree_1_dc','injury_dgree_2_dc', 
                  'legaldong_name','lrg_violt_1_dc','occrrnc_time_dc',
                  'rdse_sttus_dc','road_div','road_stle_dc','route_nm',
                   'sep_cnt', 'sexdstn_div_1_dc','sexdstn_div_2_dc','slp_cnt',
                   'tmzon_div_1_dc','wether_sttus_dc',
                  'wrngdo_vhcle_asort_dc','x_crdnt','y_crdnt'
                  ]]

In [5]:
acci_df.shape

(108398, 35)

In [ ]:
display(acci_df.head(2))

acdnt_age_1_code  acdnt_age_2_code    acdnt_dc    acdnt_dd_dc acdnt_gae_dc  \
0                39              40.0  차대차 - 측면충돌  2021년  1월  1일         중상사고   
1                40              49.0    차대차 - 기타  2021년  1월  1일         중상사고   

  acdnt_hdc acdnt_mdc          acdnt_no      acdnt_pos  acdnt_year  ...  \
0       차대차      측면충돌  2021010100100001  빽다방 사거리 앞 은평로        2021  ...   
1       차대차        기타  2021010100100002            NaN        2021  ...   

  route_nm sep_cnt sexdstn_div_1_dc sexdstn_div_2_dc  slp_cnt tmzon_div_1_dc  \
0      은평로       1                남                남        0             야간   
1      NaN       1                남                남        3             야간   

   wether_sttus_dc wrngdo_vhcle_asort_dc x_crdnt  y_crdnt  
0               맑음                    승용  948902  1955807  
1               맑음                    이륜  947601  1946291  

[2 rows x 35 columns]

In [ ]:
acci_df.columns

Index(['acdnt_age_1_code', 'acdnt_age_2_code', 'acdnt_dc', 'acdnt_dd_dc',
       'acdnt_gae_dc', 'acdnt_hdc', 'acdnt_mdc', 'acdnt_no', 'acdnt_pos',
       'acdnt_year', 'bdy_injury_part_1_dc', 'bdy_injury_part_2_dc', 'dfk_dc',
       'dmge_vhcle_asort_dc', 'dprs_cnt', 'hhdgw_at', 'inj_aplcnt_cnt',
       'injury_dgree_1_dc', 'injury_dgree_2_dc', 'legaldong_name',
       'lrg_violt_1_dc', 'occrrnc_time_dc', 'rdse_sttus_dc', 'road_div',
       'road_stle_dc', 'route_nm', 'sep_cnt', 'sexdstn_div_1_dc',
       'sexdstn_div_2_dc', 'slp_cnt', 'tmzon_div_1_dc', 'wether_sttus_dc',
       'wrngdo_vhcle_asort_dc', 'x_crdnt', 'y_crdnt'],
      dtype='object')

In [ ]:
# acci_df.to_csv('교통사고데이터_1차정제.csv', encoding='cp949')

# 2. 컬럼 의미

- acdnt_age_1_code : 가해 운전자 연령
- acdnt_age_2_code : 피해 운전자 연령
- acdnt_dc : 사고 유형
- acdnt_dd_dc : 사고 일시
- acdnt_gae_dc : 사고 내용
- acdnt_hdc : 사고 유형 중 앞부분
- acdnt_mdc : 사고 유형 중 뒷부분
- acdnt_no : 사고 번호
- acdnt_pos : 정확한 사고 위치
- acdnt_year : 사고 발생 연도
- bdy_injury_part_1_dc : 가해 운전자 상해 위치
- bdy_injury_part_2_dc : 피해 운전자 상해 위치
- dfk_dc : 사건발생 요일
- dmge_vhcle_asort_dc : 피해 운전자 차종
- dprs_cnt : 사망자 수
- hhdgw_at
- inj_aplcnt_cnt
- injury_dgree_1_dc : 가해 운전자 상해정도
- injury_dgree_2_dc : 피해 운전자 상해정도
- legaldong_name : 사건발생동(법정동명)
- lrg_violt_1_dc : 위반한 법규
- occrrnc_time_dc : 사건발생 시각
- rdse_sttus_dc : 노면 상태
- road_div
- road_stle_dc : 도로 형태
- route_nm : 사건발생 위치(도로명)
- sep_cnt
- sexdstn_div_1_dc : 가해 운전자 성별
- sexdstn_div_2_dc : 피해 운전자 성별
- slp_cnt
- tmzon_div_1_dc : 사건 당시 시간
- wether_sttus_dc : 사건 당시 날씨
- wrngdo_vhcle_asort_dc : 가해 운전자 차종
- x_crdnt : 사고위치 X좌표(5179 좌표계)
- y_crdnt : 사고위치 Y좌표(5179 좌표계)

# 3. EDA

In [ ]:
!pip install -U pandas-profiling

In [20]:
import pandas as pd
import numpy as np
import pandas_profiling

In [ ]:
acci_df.shape

(33913, 35)

In [ ]:
report = acci_df.profile_report()
report

Output hidden; open in https://colab.research.google.com to view.

In [ ]:
report.to_file('acdnt_EDA_report.html')

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

In [ ]:
acci_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33913 entries, 0 to 33912
Data columns (total 35 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   acdnt_age_1_code       33913 non-null  int64  
 1   acdnt_age_2_code       32999 non-null  float64
 2   acdnt_dc               33913 non-null  object 
 3   acdnt_dd_dc            33913 non-null  object 
 4   acdnt_gae_dc           33913 non-null  object 
 5   acdnt_hdc              33913 non-null  object 
 6   acdnt_mdc              33913 non-null  object 
 7   acdnt_no               33913 non-null  int64  
 8   acdnt_pos              21926 non-null  object 
 9   acdnt_year             33913 non-null  int64  
 10  bdy_injury_part_1_dc   33910 non-null  object 
 11  bdy_injury_part_2_dc   33910 non-null  object 
 12  dfk_dc                 33913 non-null  object 
 13  dmge_vhcle_asort_dc    33001 non-null  object 
 14  dprs_cnt               33913 non-null  int64  
 15  hh

## 결측치 확인 및 채우기

- 빈 값이 24315개, 전체의 2퍼센트
- 가해 및 피해 운전자 연령에 미분류 값이 존재
- 결측 존재 컬럼
  - acdnt_age_2_dc(피해 운전자 연령) :  912개의 NaN존재 -> 공작물충돌, 전도전복 등 사유로 차량 단독 사고 경우 존재 -> '없음'으로 대체
    - dmge_vhcle_asort_dc(피해 운전자 차종)
      - acdnt_age_2_dc와 마찬가지 사유로 NaN 존재 -> '없음'으로 대체
    - injury_dgree_2_dc(피해 운전자 상해정도)
      - acdnt_age_2_dc와 마찬가지 사유로 NaN 존재 -> '없음'으로 대체
    - sexdstn_div_2_dc(피해 운전자 성별)
      - acdnt_age_2_dc와 마찬가지 사유로 NaN 존재 -> '없음'으로 대체
  - acdnt_pos(정확한 사고 위치) : 11987개(전체 35.3%) 빈 값 존재 -> 빈값을 '기록없음'으로 대체
  - route_nm(사건발생 위치:도로명) : 4336(전체 12.8%)의 결측치 -> '기록없음'로 대체
    - road_div : 4336(전체 12.8%)의 결측치 -> route_nm에 결측이 있으면 road_div에도 결측이 있다 -> 결측을 같은 값으로 채워줌
- bdy_injury_part_1_dc, bdy_injury_part_2_dc는 3개의 결측이 존재 -> 상해 부위가 존재하지 않는 것이 아닌 기록이 되지 않은 것을로 보임
  - 상해 정도에는 중상으로 들어가 있으나, bdy_injury_part가 결측치인 것을 통해 추측
  - 양 쪽 모두 '기록없음'으로 결측치를 기록


In [6]:
# acdnt_age_2_dc(피해 운전자 연령)에 NaN이 값이 들어있는 컬럼 확인
# -> 공작물충돌, 전도전복 등 차량 단독 사고인 경우 존재 -> 빈값을 '없음'으로 대체
# dmge_vhcle_asort_dc, injury_dgree_2_dc, sexdstn_div_2_dc도 마찬가지
acci_df['acdnt_age_2_code'] = acci_df['acdnt_age_2_code'].fillna(1000)
acci_df['dmge_vhcle_asort_dc'] = acci_df['dmge_vhcle_asort_dc'].fillna('없음')
acci_df['injury_dgree_2_dc'] = acci_df['injury_dgree_2_dc'].fillna('없음')
acci_df['sexdstn_div_2_dc'] = acci_df['sexdstn_div_2_dc'].fillna('없음')

In [7]:
# acdnt_pos(정확한 사고 위치)에 다수의 빈 값 존재 -> 빈값을 미식별로 대체
acci_df['acdnt_pos'] = acci_df['acdnt_pos'].fillna('기록없음')

In [8]:
# NaN을 가진 컬럼끼리 특별한 연관성은 안보이므로 일단 기록없음으로 대체
acci_df['route_nm'] = acci_df['route_nm'].fillna('기록없음')
acci_df['road_div'] = acci_df['road_div'].fillna('기록없음')

In [9]:
acci_df['bdy_injury_part_1_dc'] = acci_df['bdy_injury_part_1_dc'].fillna('기록없음')
acci_df['bdy_injury_part_2_dc'] = acci_df['bdy_injury_part_2_dc'].fillna('기록없음')

In [10]:
acci_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 108398 entries, 0 to 108397
Data columns (total 35 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   acdnt_age_1_code       108396 non-null  float64
 1   acdnt_age_2_code       108398 non-null  float64
 2   acdnt_dc               108398 non-null  object 
 3   acdnt_dd_dc            108398 non-null  object 
 4   acdnt_gae_dc           108398 non-null  object 
 5   acdnt_hdc              108398 non-null  object 
 6   acdnt_mdc              108398 non-null  object 
 7   acdnt_no               108398 non-null  int64  
 8   acdnt_pos              108398 non-null  object 
 9   acdnt_year             108398 non-null  int64  
 10  bdy_injury_part_1_dc   108398 non-null  object 
 11  bdy_injury_part_2_dc   108398 non-null  object 
 12  dfk_dc                 108398 non-null  object 
 13  dmge_vhcle_asort_dc    108398 non-null  object 
 14  dprs_cnt               108398 non-nu

In [15]:
acci_df.to_csv('교통사고데이터_2차정제(2019-2021).csv',index=False, encoding='cp949')

In [ ]:
acci_df.acdnt_age_1_code.unique()

array([ 39,  40,  61,  70,  57,  33,  32,  68,  41,  67,  27,  59,  35,
        28,  43,  46,  24,  49,  26,  30,  34,  58,  51,  42,  62,  44,
        36,  25,  66,  22,  16,  65,  60,  54,  18,  71,  81,  79,  45,
        31, 999,  37,  72,  56,  23,  53,  38,  63,  21,  55,  50,  64,
        74,  73,  20,  29,  48,  52,  47,  69,  19,  75,  17,  82,  78,
        77,  85,  76,  86,  84,  15,  14,  80,  83,  87,   8,  13,  11,
        12,   7,  88,   9,  10,   4,  89,  92,  91,   6,  93])

In [ ]:
report = acci_df.profile_report()
report.to_file('acdnt_EDA_report.html')

- 가해 운전자, 피해 운전자 모두 연령이 미분류로 들어가있는 경우 존재 -> 일단 999로 넣어놓음
- 또한 단독충돌사고의 경우 피해 운전자 자체가 없다.

In [ ]:
acci_df.acdnt_age_1_code.unique()

array([ 39,  40,  61,  70,  57,  33,  32,  68,  41,  67,  27,  59,  35,
        28,  43,  46,  24,  49,  26,  30,  34,  58,  51,  42,  62,  44,
        36,  25,  66,  22,  16,  65,  60,  54,  18,  71,  81,  79,  45,
        31, 999,  37,  72,  56,  23,  53,  38,  63,  21,  55,  50,  64,
        74,  73,  20,  29,  48,  52,  47,  69,  19,  75,  17,  82,  78,
        77,  85,  76,  86,  84,  15,  14,  80,  83,  87,   8,  13,  11,
        12,   7,  88,   9,  10,   4,  89,  92,  91,   6,  93])